File: DSA/DevvratKhandal_590014906/Day_18/Question_1.java
import java.util.Scanner;

class ListNode {
    int val;
    ListNode next;

    ListNode(int val) {
        this.val = val;
    }
}

public class Question_1 {

    // Function to rotate linked list
    public static ListNode rotateRight(ListNode head, int k) {

        if (head == null || head.next == null || k == 0)
            return head;

        int length = 1;
        ListNode tail = head;

        while (tail.next != null) {
            tail = tail.next;
            length++;
        }

        k = k % length;

        if (k == 0)
            return head;

        // Make circular
        tail.next = head;

        int steps = length - k;
        ListNode newTail = head;

        for (int i = 1; i < steps; i++) {
            newTail = newTail.next;
        }

        ListNode newHead = newTail.next;
        newTail.next = null;

        return newHead;
    }

    // Function to print linked list
    public static void printList(ListNode head) {
        while (head != null) {
            System.out.print(head.val + " ");
            head = head.next;
        }
    }

    public static void main(String[] args) {

        Scanner sc = new Scanner(System.in);

        System.out.print("Enter number of nodes: ");
        int n = sc.nextInt();

        System.out.println("Enter linked list elements:");
        ListNode head = null, tail = null;

        for (int i = 0; i < n; i++) {
            int data = sc.nextInt();
            ListNode newNode = new ListNode(data);

            if (head == null) {
                head = newNode;
                tail = newNode;
            } else {
                tail.next = newNode;
                tail = newNode;
            }
        }

        System.out.print("Enter value of k: ");
        int k = sc.nextInt();

        head = rotateRight(head, k);

        System.out.println("Linked List after rotation:");
        printList(head);

        sc.close();
    }
}

File: DSA/Vaishnavi_590017838/Day_18/Question1.cpp
#include <iostream>
using namespace std;

struct Node {
    int data;
    Node* next;
    Node(int value) {
        data = value;
        next = nullptr;
    }
};

Node* rotateRight(Node* head, int k) {
    if(head == nullptr || head->next == nullptr || k == 0)
        return head;
    int length = 1;
    Node* tail = head;
    while(tail->next != nullptr) {
        tail = tail->next;
        length++;
    }
    k %= length;
    if(k == 0)
        return head;

    tail->next = head;
    int steps = length - k;
    Node* newTail = head;
    for(int i = 1; i < steps; i++)
        newTail = newTail->next;
    Node* newHead = newTail->next;
    newTail->next = nullptr;
    return newHead;
}
void printList(Node* head) {
    while(head != nullptr) {
        cout << head->data;
        if(head->next != nullptr)
            cout << " -> ";
        head = head->next;
    }
    cout << endl;
}
int main() {
    Node* head = new Node(1);
    head->next = new Node(2);
    head->next->next = new Node(3);
    head->next->next->next = new Node(4);
    head->next->next->next->next = new Node(5);
    int k = 2;
    head = rotateRight(head, k);
    printList(head);
    return 0;
}
